### PR TITLE
add table of contents to the specification #114

### LINKED ISSUE #114: Add table of contents to the specification
Hi!

Currently, the spec is a bit hard to read because there is no ToC which lists all methods/notifications and types. It might be a good idea to provide one! Take a look at the index at the bottom of the Dart server spec, for example, it is really convenient: https://htmlpreview.github.io/?https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/doc/api.html#Index.

### PR DESCRIPTION
**Pull request by vladdu**
resolves #114

The TOC also has icons added to specify if the messages are requests or notifications and in which directions they go. These could also be added to the descriptions.

### CODE REVIEW

File: protocol.md [hunk 1]
```diff
@@ -1640,3 +1640,47 @@ interface RenameParams {
...
+
+* [window/showMessage](#window_showMessage) notification :arrow_left:
+* [window/showMessageRequest](#window_showMessageRequest) :arrow_right_hook:
+* [window/logMessage](#window_logMessage) :arrow_left:
+* [telemetry/event](#telemetry_event) :arrow_left:
+* [textDocument/publishDiagnostics](#textDocument_publishDiagnostics) :arrow_left:
+
+Notifications client &rarr; server
+
+* [workspace/didChangeConfiguration](#workspace_didChangeConfiguration) :arrow_right:
```
Comment (kaloyan-raev) [inline protocol.md]: This link does not work. I think you miss the anchor.

File: protocol.md [hunk 2]
```diff
@@ -1640,3 +1640,47 @@ interface RenameParams {
...
+* [workspace/didChangeWatchedFiles](#workspace_didChangeWatchedFiles) :arrow_right:
+* [workspace/symbol](#workspace_symbol) :arrow_right:
```
Comment (kaloyan-raev) [inline protocol.md]: This is a request. It should be in the "Requests" section and the icon should be changed.

File: protocol.md [hunk 3]
```diff
@@ -1640,3 +1640,47 @@ interface RenameParams {
...
+
+* [initialize](#initialize) :leftwards_arrow_with_hook:
+* [shutdown](#shutdown) :leftwards_arrow_with_hook:
+* [exit](#exit) notification :arrow_right:
+* [$/cancelRequest](#cancelRequest) :arrow_right:
+
+Notifications server &rarr; client
+
+* [window/showMessage](#window_showMessage) notification :arrow_left:
+* [window/showMessageRequest](#window_showMessageRequest) :arrow_right_hook:
```
Comment (kaloyan-raev) [inline protocol.md]: This is a request, not a notification.

**Comment by vladdu** [comment]
Amended with the suggested changes. 

Changed the icons at the beginning, which gives a better feeling and grouped by area (workspace, window, textDocument). I feel it would be even better if there were different colors for the arrows, to be able to differentiate between c->s vs s->c. Maybe using real images instead of emojis is better, because the emojis are a github feature?

File: protocol.md [hunk 4]
```diff
@@ -750,15 +750,15 @@ interface LogMessageParams {
...
+#### <a name="telemetry_event"></a>Telemetry Notification
 
 >**New:** The telemetry notification is sent from the server to the client to ask the client to log a telemetry event.
 
 _Notification_:
 * method: 'telemetry/event'
 * params: 'any'
 
-#### DidChangeConfiguration Notification
+#### <a name="workspace_didChangeNotification"></a>DidChangeConfiguration Notification
```
Comment (kaloyan-raev) [inline protocol.md]: Did you mean "workspace_didChangeConfiguration" here?

**Comment by kaloyan-raev** [comment]
The new layout looks better.

I am personally fine with emojis. I read the protocol document directly on GitHub.

File: protocol.md (hunk 4 above)
Comment (vladdu) [inline protocol.md]: Silly me... Fixed.

**Comment by vladdu** [comment]
Hopefully the specification document will get generated from the schema, once #25 is implemented. The emoji names have no relation to their meaning.

**Comment by felixfbecker** [comment]
I love this, but the ToC should be at the top, not the bottom

**Comment by vladdu** [comment]
I put it at the end because it's a rather long list. The best solution would have been to have it as an aside, and with smaller text, but I think that would only be possible with raw HTML and that would make it a bit more difficult to maintain manually. 

Maybe at the top we could add a link to the TOC located at the bottom?

**Comment by felixfbecker** [comment]
Well [GitHub supports `<details>`](https://gist.github.com/ericclemmons/b146fe5da72ca1f706b2ef72a20ac39d), not sure about `<aside>`. But I still think having the ToC at the top would make it a lot more readable. I always have to use <kbd>Ctrl</kbd>+<kbd>F</kbd> to search for requests atm, and at the bottom the ToC is kind of useless.

**Comment by vladdu** [comment]
If moving the TOC at the top, do you think it's best at the very top (before "Base Protocol") or before " Actual Protocol"?

**Comment by felixfbecker** [comment]
The very top, after the introduction.

**Comment by vladdu** [comment]
Ok, moved it.

**Comment by dbaeumer** [comment]
Cool.

**Pull request merged**